src/block/Hopper.php

```diff
@@ -92,7 +97,68 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
 	}
 
 	public function onScheduledUpdate() : void{
-		//TODO
+		$world = $this->position->getWorld();
+
+		$tile = $this->position->getWorld()->getTile($this->position);
+		if(!$tile instanceof TileHopper) { $world->scheduleDelayedBlockUpdate($this->position, 1); return; }
+
+		$facingBlock = $this->getSide($this->facing);
+		$topBlock = $this->getSide(Facing::UP);
+		if(!$facingBlock instanceof HopperInteractable && !$topBlock instanceof Chest && !$topBlock instanceof Hopper) { $world->scheduleDelayedBlockUpdate($this->position, 1); return; }
+
+		$pullSuccess = $this->pull($tile);
```

Comment:
The `pull()` method should simply be implemented into the `HopperInteractable` interface. Especially since hoppers can also pull from most of the blocks they can push to.